akka-cluster-sharding-typed/src/main/scala/akka/cluster/sharding/typed/internal/ClusterActorSetImpl.scala

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (C) 2009-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.cluster.sharding.typed.internal
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import akka.actor.DeadLetterSuppression
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.Behavior
+import akka.actor.typed.Terminated
+import akka.actor.typed.scaladsl.Behaviors
+import akka.annotation.InternalApi
+import akka.cluster.sharding.ShardRegion.EntityId
+import akka.cluster.sharding.typed.ClusterShardingSettings
+import akka.cluster.sharding.typed.ClusterShardingSettings.StateStoreModeDData
+import akka.cluster.sharding.typed.ShardingEnvelope
+import akka.cluster.sharding.typed.internal.ClusterActorSetImpl.EntityParent
+import akka.cluster.sharding.typed.internal.ClusterActorSetImpl.KeepAlivePinger
+import akka.cluster.sharding.typed.scaladsl.ClusterActorSet
+import akka.cluster.sharding.typed.scaladsl.ClusterActorSetSettings
+import akka.cluster.sharding.typed.scaladsl.ClusterSharding
+import akka.cluster.sharding.typed.scaladsl.Entity
+import akka.cluster.sharding.typed.scaladsl.EntityContext
+import akka.cluster.sharding.typed.scaladsl.EntityTypeKey
+
+import scala.concurrent.duration.Duration
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[akka] object ClusterActorSetImpl {
+  object EntityParent {
+    sealed trait Command
+    case object Ping extends Command with DeadLetterSuppression
+    case object Stop extends Command
+
+    def apply[T](
+        entityContext: EntityContext[Command],
+        factory: EntityId => Behavior[T],
+        childStopMessage: T): Behavior[Command] =
+      Behaviors.setup { context =>
+        var child: Option[ActorRef[T]] = None
+
+        def startChild(): Unit = {
+          val startedChild = context.spawn(factory(entityContext.entityId), "entity")
+          context.watch(startedChild)
+          child = Some(startedChild)
+        }
+
+        context.log.debug(s"Starting ClusterActorSet actor ${entityContext.entityId}")
+        startChild()
+
+        Behaviors
+          .receiveMessage[Command] {
+            case Stop =>
+              child match {
+                case Some(ref) =>
+                  ref ! childStopMessage
+                  waitingForChildToStop()
+                case None =>
+                  Behaviors.stopped
+              }
+
+            case Ping =>
+              Behaviors.same
+          }
+          .receiveSignal {
+            case (_, Terminated(_)) =>
+              // if the child stopped manually, but we are not being re-located to another node,
+              // we can fast-restart it instead of waiting for a ping
```

Comment:
Not convinced this is a good idea. Should be possible to use ordinary supervision if it's restarted/stopped unexpectedly.

Reply:
maybe keep this PR simple, and add such when we add such manager capabilities?

Reply:
I'm fine with removing the auto-restart here.

I think that the kind of stop start @renatocaval describes should be a user concern and be a part of the actor they run in their cluster set.